c-dependencies/js-compute-runtime/builtins/secret-store.cpp

```diff
@@ -0,0 +1,197 @@
+#include "secret-store.h"
+#include "host_call.h"
+
+namespace builtins {
+
+fastly_secret_handle_t SecretStoreEntry::secret_handle(JSObject *obj) {
+  JS::Value val = JS::GetReservedSlot(obj, SecretStoreEntry::Slots::Handle);
+  return static_cast<fastly_secret_handle_t>(val.toInt32());
+}
+
+bool SecretStoreEntry::plaintext(JSContext *cx, unsigned argc, JS::Value *vp) {
+  METHOD_HEADER(0)
+  fastly_option_string_t ret;
+  fastly_error_t err;
+  // Ensure that we throw an exception for all unexpected host errors.
+  if (!xqd_fastly_secret_store_plaintext(SecretStoreEntry::secret_handle(self), &ret, &err)) {
+    HANDLE_ERROR(cx, err);
+    return false;
+  }
+
+  // None indicates the key wasn't found, so we return null.
+  if (!ret.is_some) {
+    args.rval().setNull();
+    return true;
+  }
+
+  JS::RootedString text(cx, JS_NewStringCopyUTF8N(cx, JS::UTF8Chars(ret.val.ptr, ret.val.len)));
+  JS_free(cx, ret.val.ptr);
+  if (!text)
+    return false;
+
+  args.rval().setString(text);
+  return true;
+}
+
+const JSFunctionSpec SecretStoreEntry::methods[] = {
+    JS_FN("plaintext", plaintext, 0, JSPROP_ENUMERATE), JS_FS_END};
+
+const JSPropertySpec SecretStoreEntry::properties[] = {JS_PS_END};
+
+bool SecretStoreEntry::constructor(JSContext *cx, unsigned argc, JS::Value *vp) {
+  JS_ReportErrorUTF8(cx, "SecretStoreEntry can't be instantiated directly");
+  return false;
+}
+
+JSObject *SecretStoreEntry::create(JSContext *cx, fastly_secret_handle_t handle) {
+  JS::RootedObject entry(
+      cx, JS_NewObjectWithGivenProto(cx, &SecretStoreEntry::class_, SecretStoreEntry::proto_obj));
+  if (!entry) {
+    return nullptr;
+  }
+
+  JS::SetReservedSlot(entry, Slots::Handle, JS::Int32Value(handle));
+
+  return entry;
+}
+
+bool SecretStoreEntry::init_class(JSContext *cx, JS::HandleObject global) {
+  return BuiltinImpl<SecretStoreEntry>::init_class_impl(cx, global);
+}
+
+fastly_secret_store_handle_t SecretStore::secret_store_handle(JSObject *obj) {
+  JS::Value val = JS::GetReservedSlot(obj, SecretStore::Slots::Handle);
+  return static_cast<fastly_secret_store_handle_t>(val.toInt32());
+}
+
+bool SecretStore::get(JSContext *cx, unsigned argc, JS::Value *vp) {
+  METHOD_HEADER(1)
+
+  xqd_world_string_t key_str;
+  JS::UniqueChars key = encode(cx, args[0], &key_str.len);
+  if (!key) {
+    return false;
+  }
+  // If the converted string has a length of 0 then we throw an Error
+  // because keys have to be at-least 1 character.
+  if (!key || key_str.len == 0) {
+    JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr, JSMSG_SECRET_STORE_KEY_EMPTY);
+    return false;
+  }
+  key_str.ptr = key.get();
```

Comment:
What do you think about reworking this so that you only track the size of the encoded string? Eagerly declaring the `key_str` was a little surprising for me, as I had assumed that we would be using it right away. Moving the declaration and initialization down to where it's used would help to clarify why it's ok that it's holding onto memory managed by a `UniqueChars`.